client/src/components/ProtectedRoute/ProtectedRoute.tsx

```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import { RouteComponentProps } from 'react-router';
+import { Route, Redirect, RouteProps } from 'react-router-dom';
+import { useAuth } from '../../context/useAuthContext';
+
+interface ProtectedRoute extends RouteProps {
+  component: React.ComponentType<RouteComponentProps<any>> | React.ComponentType<any> | undefined;
+}
+
+export default function ProtectedRoute({ component: Component, ...rest }: ProtectedRoute) {
+  const { loggedInUser } = useAuth();
+  const redirectPath = '/login';
+
+  if (!Component) return null;
+
+  return (
+    <Route
+      {...rest}
+      render={(props) => {
+        if (loggedInUser) {
+          return <Component {...props} />;
+        } else {
+          return (
+            <Redirect
+              to={{
+                pathname: redirectPath,
+                state: {
+                  from: props.location,
+                },
+              }}
+            />
+          );
+        }
+      }}
+    />
```

Comment:
This is very good overall 💯 

`Route` components have [a few different ways to render the contents](https://v5.reactrouter.com/web/api/Route/route-render-methods) and ideally, we should keep that same for `ProtectedRoute` to provide better development experience. Luckily, achieving this is fairly simple - all we have to is forward all of the props we receive to a `Route` component and it will handle rendering for us.